### ISSUE TITLE
android.database.sqlite.SQLiteCantOpenDatabaseException

### ISSUE THREAD

**Issue by rameshvoltella** [OneSignal/OneSignal-Android-SDK#945]
Hi,

in my app i am getting  SQLiteCantOpenDatabaseException for lava device

Error log
```
android.database.sqlite.SQLiteCantOpenDatabaseException: 
 
  at android.database.sqlite.SQLiteConnection.nativeOpen (SQLiteConnection.java)
 
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:214)
 
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:198)
 
  at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked (SQLiteConnectionPool.java:463)
 
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:185)
 
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:177)
 
  at android.database.sqlite.SQLiteDatabase.openInner (SQLiteDatabase.java:838)
 
  at android.database.sqlite.SQLiteDatabase.open (SQLiteDatabase.java:823)
 
  at android.database.sqlite.SQLiteDatabase.openDatabase (SQLiteDatabase.java:716)
 
  at android.app.ContextImpl.openOrCreateDatabase (ContextImpl.java:669)
 
  at android.content.ContextWrapper.openOrCreateDatabase (ContextWrapper.java:289)
 
  at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked (SQLiteOpenHelper.java:223)
 
  at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase (SQLiteOpenHelper.java:163)
 
  at com.onesignal.OneSignalDbHelper.getWritableDbWithRetries (OneSignalDbHelper.java)
 
  at com.onesignal.OneSignalDbHelper.upgradeOutcomeTableRevision1To2 (OneSignalDbHelper.java)
 
  at com.onesignal.OneSignalCacheCleaner$1.run (OneSignalCacheCleaner.java)
 
  at java.lang.Thread.run (Thread.java:761)
```

More info about the crash 
Device - Lava Z60 (Z60)
Android version- Android 7.0
Onesignal version- 3.12.5

i have seen the new release of onesignal 3.12.6 but in the release not i haven't seen anything mentioned about this crash please help the crash is increasing.

**Comment by mikechoch** [comment]
A little more detail would be nice. 
1. When does this occur? Do you have steps to reproduce?
2. What version of OneSignal did you notice this start happening?

**Comment by rameshvoltella** [comment]
1)Dont know both
2)its happening from 3.12.5 i think
i got this crash today also in iBall iBall Slide Snap 4G2 (Snap_4G2), Android 7.0
So i guess its not happening only in lava
The log i got in i ball

```
android.database.sqlite.SQLiteCantOpenDatabaseException: 
 
  at android.database.sqlite.SQLiteConnection.nativeOpen (SQLiteConnection.java)
 
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:214)
 
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:198)
 
  at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked (SQLiteConnectionPool.java:463)
 
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:185)
 
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:177)
 
  at android.database.sqlite.SQLiteDatabase.openInner (SQLiteDatabase.java:838)
 
  at android.database.sqlite.SQLiteDatabase.open (SQLiteDatabase.java:823)
 
  at android.database.sqlite.SQLiteDatabase.openDatabase (SQLiteDatabase.java:716)
 
  at android.app.ContextImpl.openOrCreateDatabase (ContextImpl.java:664)
 
  at android.content.ContextWrapper.openOrCreateDatabase (ContextWrapper.java:289)
 
  at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked (SQLiteOpenHelper.java:223)
 
  at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase (SQLiteOpenHelper.java:163)
 
  at com.onesignal.OneSignalDbHelper.getWritableDbWithRetries (OneSignalDbHelper.java)
 
  at com.onesignal.OneSignalDbHelper.upgradeOutcomeTableRevision1To2 (OneSignalDbHelper.java)
 
  at com.onesignal.OneSignalCacheCleaner$1.run (OneSignalCacheCleaner.java)
 
  at java.lang.Thread.run (Thread.java:761)
```

**Comment by jkasten2** [comment]
`SQLiteCantOpenDatabaseException` is normally a permission issue if trying to use external storage, however the OneSignal DB is stored with the app data so no additional permission is needed.

My best guess is that the devices are out of space if you only are seeing a hand full of these crashes.

**Comment by rameshvoltella** [comment]
@jkasten2 So what should we do to solve this issue?Is onesignal Storing lot of data in sqlite?

**Comment by jkasten2** [comment]
@rameshvoltella The limited space issue is a best guess but haven't confirmed the exact error you would get if the device as out of space. The OneSignal SDK should only be taking ~1MB for it's SQLite DB, however this could be higher if the device can get a very high volume of notifications. Only 1 week of notifications are stored on the device by the OneSignal SDK.

**Comment by dmacli** [comment]
We have an app with 4 million users a day and this is happening about 1,000 times a day.

`unknown error (code 14): Could not open database
android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 14): Could not open database
	at android.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:210)
	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:194)
	at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:464)
	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:186)
	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:178)
	at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:918)
	at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:895)
	at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:798)
	at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:654)
	at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:289)
	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:223)
	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
	at com.onesignal.bf.a(SourceFile:125)
	at com.onesignal.bc$1.run(SourceFile:26)
	at java.lang.Thread.run(Thread.java:761)`

**Comment by dmacli** [comment]
Also got another report that looks like this:

`
unknown error (code 1806): Could not open database
#################################################################
Error Code : 1806 (SQLITE_CANTOPEN_EACCES)
Caused By : Application has no permission to open the specified database file.
	(unknown error (code 1806): Could not open database)
#################################################################
android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 1806): Could not open database
#################################################################
Error Code : 1806 (SQLITE_CANTOPEN_EACCES)
Caused By : Application has no permission to open the specified database file.
	(unknown error (code 1806): Could not open database)
#################################################################
	at android.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:262)
	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:205)
	at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:649)
	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:255)
	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:222)
	at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:1191)
	at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:1146)
	at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:861)
	at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:729)
	at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:310)
	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:254)
	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
	at com.onesignal.bf.a(SourceFile:125)
	at com.onesignal.bc$1.run(SourceFile:26)
	at java.lang.Thread.run(Thread.java:764)
`

**Comment by rgomezp** [comment]
Howdy,
We are still investigating this issue. Thanks for your patience

**Comment by dmacli** [comment]
It seems to have gotten better after 3.13.0.  Much, much less frequent.

**Comment by jkasten2** [comment]
We have fixed the issue in PR #995 with the help of these details and was just released today in our [3.12.2 release](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/3.13.2)

**Comment by asadi110** [comment]
It Has occurred in sdk 3.13.2 : com.onesignal.OneSignalDbHelper.getWritableDbWithRetries(OneSignalDbHelper.java:137)
::
Fatal Exception: android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): , while compiling: PRAGMA journal_mode
at android.database.sqlite.SQLiteConnection.nativePrepareStatement(SQLiteConnection.java)
at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
at android.database.sqlite.SQLiteConnection.executeForString(SQLiteConnection.java:634)
at android.database.sqlite.SQLiteConnection.setJournalMode(SQLiteConnection.java:320)
at android.database.sqlite.SQLiteConnection.setWalModeFromConfiguration(SQLiteConnection.java:294)
at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:215)
at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:193)
at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:463)
at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:185)
at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:177)
at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:806)
at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:791)
at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:694)
at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:1155)
at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:268)
at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:223)
at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
at com.onesignal.OneSignalDbHelper.getWritableDbWithRetries(OneSignalDbHelper.java:137)
at com.onesignal.OneSignalCacheCleaner$1.run(OneSignalCacheCleaner.java:26)
at java.lang.Thread.run(Thread.java:818)
